llvm/lib/SYCLLowerIR/SYCLJointMatrixTransform.cpp

```diff
@@ -60,34 +127,61 @@ bool transformAccessChain(Function *F) {
     // from sycl::joint_matrix class object if it's used in __spirv_AccessChain
     // function call. It's necessary because otherwise OpAccessChain indices
     // would be wrong.
-    Instruction *Ptr =
-        dyn_cast<Instruction>(CI->getArgOperand(0)->stripPointerCasts());
+    Instruction *Ptr = dyn_cast<Instruction>(
+        stripPointerCastsAndOffsets(CI->getArgOperand(0)));
     if (!Ptr || !isa<AllocaInst>(Ptr))
       continue;
-    StructType *WrapperMatrixTy =
-        dyn_cast<StructType>(cast<AllocaInst>(Ptr)->getAllocatedType());
-    if (!WrapperMatrixTy)
-      continue;
-    TargetExtType *MatrixTy =
-        dyn_cast<TargetExtType>(WrapperMatrixTy->getElementType(0));
-    if (!MatrixTy)
+
+    Type *AllocaTy = cast<AllocaInst>(Ptr)->getAllocatedType();
+    // It may happen that sycl::joint_matrix class object is wrapped into
+    // nested arrays. We need to find the innermost type to extract
+    if (StructType *WrapperMatrixTy =
+            dyn_cast<StructType>(getInnermostType(AllocaTy))) {
+      TargetExtType *MatrixTy = extractMatrixType(WrapperMatrixTy);
+      if (!MatrixTy)
+        continue;
+
+      AllocaInst *Alloca = nullptr;
+      {
+        IRBuilder Builder(CI);
+        IRBuilderBase::InsertPointGuard IG(Builder);
+        Builder.SetInsertPointPastAllocas(CI->getFunction());
+        Alloca = Builder.CreateAlloca(replaceInnermostType(AllocaTy, MatrixTy));
+        Alloca->takeName(Ptr);
+      }
+      Ptr->replaceAllUsesWith(Alloca);
+      Ptr->dropAllReferences();
+      Ptr->eraseFromParent();
+      ModuleChanged = true;
+    }
+
+    // In case spirv.CooperativeMatrixKHR is used in arrays, we also need to
+    // insert GEP to get pointer to target exention type and use it instead of
+    // pointer to sycl::joint_matrix class object when it is passed to
+    // __spirv_AccessChain
+    // First we check if the argument came from a GEP instruction
+    GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(
+        stripPointerCastsAndOffsets(CI->getArgOperand(0), true));
```

Comment:
nit: can we add an inline comment explaining what the `true` is, like `/*StopOnGEPs=*/true`

Reply:
done